pkg/osvscanner/customgitignore/walk_up_to_root.go

```diff
@@ -0,0 +1,234 @@
+	"github.com/go-git/go-git/v5"
+	"github.com/go-git/go-git/v5/plumbing/format/gitignore"
+)
+
+// Takes a dir and processes .gitignore files from it.
+//
+// This uses go-git under the hood and returns a slice
+// of go-git's gitignore.Pattern structs.
+//
+// The actual parsing is inteded to be similar to how tools
+// rg work, but means that `path` may not necessarily be
+// the root of a git repo, and can produces these parsing
+// behaviours:
+//
+// - parsing file:
+// 	- find the file's dir and then run the following rules...
+//
+// - parsing dir at the root of a git-repo
+// 	- with recursive flag:
+// 		- read all .gitignore files in repo
+// 		- read .git/info/exclude for repo
+// 	- without recursive flag:
+// 		- read .gitignore files in repo-root
+// 		- read .git/info/exclude for repo
+//
+// - parsing dir within, a git-repo
+// 	- with recursive flag:
+// 		- read .gitignore in start dir
+// 		- read all .gitignore files in child-dirs
+// 		- read all .gitignore files in parent dirs up to and including repo-root
+// 			(ie only dirs that are an ancestor)
+// 		- read .git/info/exclude for repo
+//  - NOTE: the dir you're passing in directly could be a dir that is ignored
+//      by a parent's .gitignore or the per-repo exclude file; in this
+//      case, we still process the dir's .gitignore file and, possibly, but
+//      not it's sub-dirs
+//
+// 	- without recursive flag:
+// 		- read .gitignore in start dir
+// 		- read all .gitignore files in parent dirs up to and including repo-root
+// 			(ie only dirs that are an ancestor)
+// 		- read .git/info/exclude for repo
+//
+// - parsing a plain dir: .gitignore files are ignored
+//
+// In all cases any dirs matched by a previously read
+// .gitignore are skipped, unless it's the path (ie directly
+// supplied by the user).
+func ParseGitIgnores(path string, recursive bool) (returnPs []gitignore.Pattern, repoRootPath string, err error) {
+	// We need to parse .gitignore files from the root of the git repo to correctly identify ignored files
+	var fs billy.Filesystem
+	var ps, newPs []gitignore.Pattern
+
+	// Normalise to path (or directory containing path if it's a file)
+	path, err = getNormalisedDir(path)
+	if err != nil {
+		return nil, "", err
+	}
+
+	repo, err := git.PlainOpenWithOptions(path, &git.PlainOpenOptions{DetectDotGit: true})
+	if err != nil && !errors.Is(err, git.ErrRepositoryNotExists) {
+		return ps, "", err
+	}
+
+	// not in a git repo; do not read .gitignore files
+	// (and ignore recursive setting)
+	if errors.Is(err, git.ErrRepositoryNotExists) {
+		return ps, path, nil
+	}
+
+	// inside a git repo
+
+	repoRootPath, err = getRepoRootPath(repo)
+	if err != nil {
+		return ps, "", err
+	}
+
+	pathAbs, err := filepath.Abs(path)
+	if err != nil {
+		return ps, "", err
+	}
+
+	// path WRT git-root
+	pathRel, err := filepath.Rel(repoRootPath, pathAbs)
+	if err != nil {
+		return ps, "", err
+	}
+
+	fs = osfs.New(repoRootPath)
+
+	// Read parent dirs
```

Comment:
@another-rex This comment is here as a "code after this is reading the parent dirs" comment, in the same way [this comment](https://github.com/google/osv-scanner/pull/797/files#diff-8b3e805815de31b538fd3c919f4a0d341ad6382fecd193cc1e4dfd38ae4c4331R90) indicates a notable change is the state of the parser.

How would you prefer this is handled?

1. remove this comment
2. remove all "state change" comments
3. change to something like `// --- Reading parent dirs ---` (or other similar style)
4. something else

Reply:
I think what confused me is that immediately underneath you are reading the current directory (which now I see can be interpreted as the "parent" directory of the file passed in). Hmm I think we can just remove this comment since you already comment the individual sections.

Reply:
any word on how you want this handled @another-rex ?

Reply:
Let's just remove this comment